### PR TITLE
Add more ZA modes

### PR DESCRIPTION
**Pull request by Leporacanthicus**
This adds support for preserving and sharing ZA Modes to the MLIR infrastructure. The functionality already exists in LLVM, so just "linking the pieces together".

### CODE REVIEW

File: mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:32]: Keep the link to the ACLE

File: mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:37]: It's a little odd to have 'PreservesZA' as a ZA mode, as it's an attribute that can be used alongside `arm_new_za` or `arm_shared_za`, not an independent mode.

File: mlir/test/Dialect/ArmSME/enable-arm-za.mlir
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/enable-arm-za.mlir:23]: That's unlikely what you meant ;-)

**Review by banach-space — COMMENTED** [review]
> This adds support for preserving and sharing ZA Modes to the MLIR infrastructure. The functionality already exists in LLVM, so just "linking the pieces together".

Mats, I'd appreciate if we could be a bit more stricter with the documentation. Could you add links with definitions of these attributes? (I guess both ACLE and LLVM IR are relevant). And could you specify which attributes are already supported and which are added in this patch? Thank you :)

**Comment by Leporacanthicus** [comment]
I have pushed an update that I think resolves the comments provided. It includes a comment explaining that the legality of any attributes used is up to the higher level user of these attributes.  I think that makes the sense.

Thanks for the comments, valuable as always.

File: mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:34]: [nit] 
```suggestion
// Various frontends (e.g. Flang) that build on top of this may restrict or enforce how these attributes
```

**Review by banach-space — APPROVED** [review]
LGTM, but please wait for @MacDue to approve before landing :)

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (MacDue) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1723]: Maybe slightly rewrite this to show new/shared are mutually exclusive, and `za_preserved` can be used with either.
This would closer match the streaming mode handling above this. 

```suggestion
  if (func->hasFnAttribute("aarch64_pstate_za_new"))
    funcOp.setArmNewZa(true);
  else if (func->hasFnAttribute("aarch64_pstate_za_shared"))
    funcOp.setArmSharedZa(true);    
  // za_preserved can be used with za_new or za_shared. 
  if (func->hasFnAttribute("aarch64_pstate_za_preserved"))
    funcOp.setArmPreservesZa(true);

```

**Review by MacDue — APPROVED** [review]
LGTM, just one thing:

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (MacDue) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1722]: Typo: eiether -> ether

**Review by c-rhodes — COMMENTED** [review]
the attributes also need adding to `mlir/lib/Target/LLVMIR/ModuleTranslation.cpp` with corresponding tests in `mlir/test/Target/LLVMIR/llvmir.mlir`

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (definelicht) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2327]: There is a `:` missing, so this is not actually checked 🙂

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (definelicht) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2334]: There is a `:` missing, so this is not actually checked 🙂

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (definelicht) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2341]: There is a `:` missing, so this is not actually checked 🙂

**Review by definelicht — COMMENTED** [review]
While modifying `llvmir.mlir` I noticed that there are some `CHECK`s here that are missing a `:` and thus aren't actually tested 😬 Could you fix this in a followup?

Please also add splits between the tests!

**Comment by Dinistro** [comment]
@definelicht This should be addressed in https://github.com/llvm/llvm-project/pull/93476

**Pull request merged**
